### ISSUE TITLE
[flang] Missing diagnostic on ambiguous generic type-bound procedure - a different case

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#131914]
Consider the following case:
```
module genericName

   type b1
      integer :: i
      contains
         procedure, nopass :: threeargs1
         generic :: threeargs => threeargs1
   end type

   type, extends(b1) :: c1
      contains
         procedure, nopass :: threeargs2
         generic :: threeargs => threeargs2
   end type


   contains

      subroutine threeargs1(x,y,z)
         real, optional, intent(in) :: x
         integer, intent(in) :: y
         real, intent(out) :: z
      end subroutine

      subroutine threeargs2(x,z,y)
         integer, optional, intent(in) :: x
         integer, intent(in) :: z
         real, intent(out) :: y
      end subroutine

end module
```

Generic procedure `threeargs1` and `threddargs2` are ambiguous.
All ifort, gfortran and XLF issue an error message.

This is a different variation from [issue #128760](https://github.com/llvm/llvm-project/issues/128760)

**Comment by klausler** [comment]
Can you write a call of the generic procedure that is ambiguous and not caught as such?

**Comment by DanielCChen** [comment]
> Compile with `-pedantic` and you will observe the message
> 
> ```
> portability: Generic 'threeargs' should not have specific procedures 'threeargs1' and 'threeargs2' as their interfaces are not distinguishable by the rules in the standard
> ```
> 
> It turns out that cases with optional or polymorphic arguments can be safe from ambiguity but fail the rules for distinguishability that appear in the standard. Since they appear in some existing code, I accept them, with this optional portability warning. If any particular generic procedure reference ever matches multiple specific procedures, a hard error will appear.

That sounds good. Thanks!
